### ISSUE TITLE
Marathon-lb keeps restarting because of a bad Health Check

### ISSUE THREAD

**Issue by amramtamir** [mesosphere/marathon-lb#504]
Hi,
I am facing an issue where the Marathon-LB is restarting every 3-5 minutes due to a bad health check.
As a workaround I have removed the health check, but I do want to figure out the issue.

Details:
DC/OS - 1.88
Marathon-LB - from universe , version : mesosphere/marathon-lb:v1.4.2
Mesos - 1.0.3
Marathon - 1.3.9

HTTP health check: 
/_haproxy_health_check

This is the log tail : 

2017-09-26 11:19:46,942 marathon_lb: received event of type health_status_changed_event
2017-09-26 11:19:46,942 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:46,942 marathon_lb: fetching apps
2017-09-26 11:19:46,948 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:46,948 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:46,948 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:46,948 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:46,948 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:49,964 marathon_lb: Connection error(None): None
2017-09-26 11:19:51,972 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,972 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,972 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,973 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,973 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,973 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,973 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,973 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,974 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,974 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,974 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,974 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,974 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,974 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,975 marathon_lb: received event of type failed_health_check_event
2017-09-26 11:19:51,975 marathon_lb: received event of type unhealthy_task_kill_event
2017-09-26 11:19:51,975 marathon_lb: received event of type failed_health_check_event
Received killTask for task marathon-lb-default.3f0c1333-a2ac-11e7-ab63-70b3d5800004

**Comment by karthiksadhasivam** [comment]
I am facing the same issue. 1 out of 3 instances are showing this message and continuosly restarting. I also see the below error message in the task failure page,

Task was killed since health check failed. Reason: BufferOverflowException: Exceeded configured max-open-requests value of [32]. This means that the request queue of this pool (HostConnectionPoolSetup(10.xxx.xxx.xxx,9090,ConnectionPoolSetup(ConnectionPoolSettings(4,0,5,32,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)),TCPTransport(None,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)))),akka.http.scaladsl.HttpConnectionContext$@43765221,akka.event.MarkerLoggingAdapter@7d9e7e43))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.

**Comment by deric** [comment]
@karthiksadhasivam the error message means that Marathon is trying to execute a HTTP health check. Through `adminrouter` an akka message to `mesos-slave` (public one in this case) is sent, which should tell the agent to execute the remote health check. For some reason the communication with Mesos Agent fails.

**Comment by vitosans** [comment]
@deric 

Tracked my problem down to being the same as this

```
123070014:Oct 23 12:20:15 master-02 marathon[5664]: [2017-10-23 12:20:15,074] #033[31mWARN #033[0;39m Performing health check for app=/marathon-lb instance=instance [marathon-lb.marathon-11d70949-b827-11e7-a077-0a648d3ab8c2] port=None failed with exception (mesosphere.marathon.core.health.impl.HealthCheckWorkerActor:marathon-akka.actor.default-dispatcher-91)
123070365-Oct 23 12:20:15 master-02 marathon[5664]: akka.stream.BufferOverflowException: Exceeded configured max-open-requests value of [32]. This means that the request queue of this pool (HostConnectionPoolSetup(192.168.1.163,9090,ConnectionPoolSetup(ConnectionPoolSettings(4,0,5,32,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)),TCPTransport(None,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)))),akka.http.scaladsl.HttpConnectionContext$@11d684d3,akka.event.MarkerLoggingAdapter@33925d55))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.
123071853-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.http.impl.engine.client.PoolInterfaceActor.<init>(PoolInterfaceActor.scala:76)
123071986-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.http.impl.engine.client.PoolInterfaceActor$$anonfun$props$1.apply(PoolInterfaceActor.scala:32)
123072135-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.http.impl.engine.client.PoolInterfaceActor$$anonfun$props$1.apply(PoolInterfaceActor.scala:32)
123072284-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.TypedCreatorFunctionConsumer.produce(IndirectActorProducer.scala:87)
123072413-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.Props.newActor(Props.scala:213)
123072505-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.ActorCell.newActor(ActorCell.scala:562)
123072605-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.ActorCell.create(ActorCell.scala:588)
123072703-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:461)
123072806-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
123072910-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
123073025-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.dispatch.Mailbox.run(Mailbox.scala:223)
123073119-Oct 23 12:20:15 master-02 marathon[5664]: #011at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
123073214-Oct 23 12:20:15 master-02 marathon[5664]: #011at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
123073332-Oct 23 12:20:15 master-02 marathon[5664]: #011at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
123073462-Oct 23 12:20:15 master-02 marathon[5664]: #011at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
123073584-Oct 23 12:20:15 master-02 marathon[5664]: #011at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
123073715-Oct 23 12:20:15 master-02 marathon[5664]: [2017-10-23 12:20:15,075] #033[34mINFO #033[0;39m Received health result for app [/marathon-lb] version [2017-10-23T19:03:00.932Z]: [Unhealthy(instance [marathon-lb.marathon-11d70949-b827-11e7-a077-0a648d3ab8c2],2017-10-23T19:03:00.932Z,BufferOverflowException: Exceeded configured max-open-requests value of [32]. This means that the request queue of this pool (HostConnectionPoolSetup(192.168.1.163,9090,ConnectionPoolSetup(ConnectionPoolSettings(4,0,5,32,1,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)),TCPTransport(None,ClientConnectionSettings(Some(User-Agent: akka-http/10.0.6),10 seconds,1 minute,512,None,<function0>,List(),ParserSettings(2048,16,64,64,8192,64,268435456,256,1048576,Strict,RFC6265,true,Full,Error,Map(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, Content-MD5 -> 0, Date -> 0, If-Match -> 0, If-None-Match -> 0, User-Agent -> 32),false,<function1>,<function1>,<function2>)))),akka.http.scaladsl.HttpConnectionContext$@11d684d3,akka.event.MarkerLoggingAdapter@33925d55))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.,2017-10-23T19:20:15.074Z,true)] (mesosphere.marathon.core.health.impl.HealthCheckActor:marathon-akka.actor.default-dispatcher-105)
```

Still not sure whats going on, this is now a fresh installed cluster with nothing running on it but two marathon-lb instances - funny that the internal lb can pass this healthcheck but the external can't.

**Comment by deric** [comment]
On `slave_public` you should be able to see something like this:
```
$ ngrep -W byline -qilwv tcp dst port 9090

T marathon.ip.address:51790 -> slave.public.ip:9090 [AP]
GET /_haproxy_health_check HTTP/1.1.
Host: slave.public.ip:9090.
User-Agent: akka-http/10.0.6.
.
```
where the port `9090` is `$PORT0` in `marathon-lb` configuration.

Make sure you're able to execute from `master02`:
```
curl -XGET slave.public.ip:9090/_haproxy_health_check
```

**Comment by drewkerrigan** [comment]
@vitosans I've been trying to investigate and reproduce this - I haven't been able to at all on CoreOS, I did see the error once on CentOS I think, still working on getting it to happen there. What DC/OS version / OS version are you using?

**Comment by vitosans** [comment]
@drewkerrigan 

I was able to figure out what was causing this. I was evaluating portworx and they have a proxy you need to load that will provide public access to their light house application: 

```json
{
    "acceptedResourceRoles": [
        "slave_public"
    ],
    "cmd": "/proxyfiles/bin/start portworx $PORT0",
    "container": {
        "docker": {
            "image": "mesosphere/repoxy:2.0.0"
        },
        "type": "DOCKER",
        "volumes": [
            {
                "containerPath": "/opt/mesosphere",
                "hostPath": "/opt/mesosphere",
                "mode": "RO"
            }
        ]
    },
    "cpus": 0.1,
    "env": {
        "PROXY_ENDPOINT_0": "Lighthouse,http,lighthouse-0-start,mesos,8085,/,/"
    },
    "id": "/repoxy",
    "instances": 1,
    "labels": {
        "HAPROXY_GROUP": "external"
    },
    "mem": 128,
    "portDefinitions": [
        {
            "port": 9998,
            "protocol": "tcp"
        },
        {
            "port": 9999,
            "protocol": "tcp"
        }
    ],
    "requirePorts": true
}
```

This is what was causing the issue. Once I stopped deploying this all is happy. Still not sure why marathon-lb wouldn't recover even after I removed this deployment. 

If you want to reproduce I think all you need to do is deploy that, and then deploy another application that uses the label 

```json
"labels": {
        "HAPROXY_GROUP": "external"
    },
```

Wait a few mins and it should fall over and it will never recover.

**Comment by vitosans** [comment]
Would love to see what the root cause/post-mortem is and how to work around such events if I come across them in the future. For whatever reason this just hosed marathon-lb ( external ) - reinstall didn't even fix it. marathon-lb(internal) was not effected at all.

**Comment by bastrich** [comment]
https://jira.mesosphere.com/browse/MARATHON-7572

**Comment by justinrlee** [comment]
There's an intermittent issue with Marathon in 1.10.x with marathon-style (`HTTP`) healthchecks, where if you have more than some number of outstanding sessions they will all fail.  Not sure on the exact root cause, but this is resolved by switching to mesos-style healthchecks (`MESOS_HTTP`)

**Comment by vespian** [comment]
I am going to close this issue. I think we have two independent problems here:
a) one with intermittent Marathon issue that @justinrlee described
b) the other one, described by @vitosans 

b) was already addressed by @jeid64  in [1]. As for a) - Marathon 1.4 deprecated Marathon-based healthchecks [1], and it is available in all our supported versions: 1.9, 1.10, 1.11 and in the master. What is more, the marathon-lb package by default uses mesos healthchecks starting with version 1.11.2. The change was made without bumping version (it is just a package configuration change), so in some cases, a reinstall or an update of the configuration may be needed.

Let me know if there is anything else I can help with. Thanks!

[1] https://jira.mesosphere.com/browse/DCOS_OSS-2750?focusedCommentId=127237&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-127237
[2] https://github.com/mesosphere/marathon/releases/tag/v1.4.0
[3] https://github.com/mesosphere/universe/pull/1611
